### ISSUE TITLE
fast::MQTT_communicator must be passed as a non-const reference?

### ISSUE THREAD

**Issue by jbreitbart** [fast-project/fast-lib#25]
I just tried to pass MQTT_communicator to a function and it looks like it must be passed as a non-const reference. Is there any specific reason for this? If I pass it as a const-reference, I cannot call get_message/send_message as these functions are not qualified with const. Call by value seems to be problematic, as the destructor closes the connection to the broker(?).

**Comment by timohl** [comment]
I would agree that those functions should be qualified with `const`.
However, the class `mosqpp::mosquittopp` of the mosquittopp library does not use `const` at all.
Thus, the method `publish()` cannot be called if `send_message()` were qualified as `const`.
Changing `mosqpp::mosquittopp` from being inherited to being a mutable member variable does not work either, since inheritance is required to override some callback functions.
Maybe, wrapping  `mosqpp::mosquittopp` in a class that provides `const`-correct access could work.

So in conclusion, I would recommend using a non-const reference.
Please tell me if this causes any problems as I will then look into a `const`-correct implementation using a wrapper class.

**Comment by jbreitbart** [comment]
Hmm, so I haven't checked the implementation, but if the design of the underlying library is broken can't we just respond with an ugly hack like (warning, code below has never seen a compiler):

```
void send_message() const {
    return (const_cast<T*>(this))->send_message();
}
```

That way the code using fast-lib would look ok-ish. Or is there anything stopping use from using this approach?

Anyway, if that is a lot of work, I am fine with leaving it as is and remove the const-ness in my code.

**Comment by timohl** [comment]
Yes, this seems possible, but is indeed kind of ugly.
However, I have just written a patch for the mosquittopp library making it const-correct.
This was not much work and the fast-lib code still looks clean.
I will test the changes and try to push it tomorrow if everything works.

**Comment by timohl** [comment]
This should now be fixed.
